# 1. Importing required libraries

In [1]:
import pandas as pd
import numpy as np
import os
from sklearn.model_selection import train_test_split

# 2. Reading the data

In [2]:
data_dir = r"Datasets"
project_dir = r"D:\Nikss\Projects\Flight Price Prediction"

In [3]:
# get_data function
def get_data(name):
    file_name = f"{name}.csv"
    path = os.path.join(project_dir, data_dir, file_name)
    df = pd.read_csv(path)
    return df

In [4]:
df = get_data('flight_price')
df

Airline Date_of_Journey    Source Destination  \
0           IndiGo      24/03/2019  Banglore   New Delhi   
1        Air India       1/05/2019   Kolkata    Banglore   
2      Jet Airways       9/06/2019     Delhi      Cochin   
3           IndiGo      12/05/2019   Kolkata    Banglore   
4           IndiGo      01/03/2019  Banglore   New Delhi   
...            ...             ...       ...         ...   
10678     Air Asia       9/04/2019   Kolkata    Banglore   
10679    Air India      27/04/2019   Kolkata    Banglore   
10680  Jet Airways      27/04/2019  Banglore       Delhi   
10681      Vistara      01/03/2019  Banglore   New Delhi   
10682    Air India       9/05/2019     Delhi      Cochin   

                       Route Dep_Time  Arrival_Time Duration Total_Stops  \
0                  BLR → DEL    22:20  01:10 22 Mar   2h 50m    non-stop   
1      CCU → IXR → BBI → BLR    05:50         13:15   7h 25m     2 stops   
2      DEL → LKO → BOM → COK    09:25  04:25 10 Jun      19h     2 stops   
3            CCU → NAG → BLR    18:05         23:30   5h 25m      1 stop   
4            BLR → NAG → DEL    16:50         21:35   4h 45m      1 stop   
...                      ...      ...           ...      ...         ...   
10678              CCU → BLR    19:55         22:25   2h 30m    non-stop   
10679              CCU → BLR    20:45         23:20   2h 35m    non-stop   
10680              BLR → DEL    08:20         11:20       3h    non-stop   
10681              BLR → DEL    11:30         14:10   2h 40m    non-stop   
10682  DEL → GOI → BOM → COK    10:55         19:15   8h 20m     2 stops   

      Additional_Info  Price  
0             No info   3897  
1             No info   7662  
2             No info  13882  
3             No info   6218  
4             No info  13302  
...               ...    ...  
10678         No info   4107  
10679         No info   4145  
10680         No info   7229  
10681         No info  12648  
10682         No info  11753  

[10683 rows x 11 columns]

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10683 entries, 0 to 10682
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   Airline          10683 non-null  object
 1   Date_of_Journey  10683 non-null  object
 2   Source           10683 non-null  object
 3   Destination      10683 non-null  object
 4   Route            10682 non-null  object
 5   Dep_Time         10683 non-null  object
 6   Arrival_Time     10683 non-null  object
 7   Duration         10683 non-null  object
 8   Total_Stops      10682 non-null  object
 9   Additional_Info  10683 non-null  object
 10  Price            10683 non-null  int64 
dtypes: int64(1), object(10)
memory usage: 918.2+ KB


In [6]:
df.dtypes

Airline            object
Date_of_Journey    object
Source             object
Destination        object
Route              object
Dep_Time           object
Arrival_Time       object
Duration           object
Total_Stops        object
Additional_Info    object
Price               int64
dtype: object

In [7]:
df.isnull().sum()

Airline            0
Date_of_Journey    0
Source             0
Destination        0
Route              1
Dep_Time           0
Arrival_Time       0
Duration           0
Total_Stops        1
Additional_Info    0
Price              0
dtype: int64

# Observations:
* Dataset contains `10,683 rows` and `11 columns`.
* It has only 2 missing values in `Route` and `Total_Stops` (One in each column).
* Datatype of some columns isn't appropriate.

In [8]:
# Changing columns' names to lower

df.columns = df.columns.str.lower()

# 3. Preliminary Analysis

# 3.1 Check Datatypes

In [9]:
df.dtypes

airline            object
date_of_journey    object
source             object
destination        object
route              object
dep_time           object
arrival_time       object
duration           object
total_stops        object
additional_info    object
price               int64
dtype: object

In [10]:
df.head()

airline date_of_journey    source destination                  route  \
0       IndiGo      24/03/2019  Banglore   New Delhi              BLR → DEL   
1    Air India       1/05/2019   Kolkata    Banglore  CCU → IXR → BBI → BLR   
2  Jet Airways       9/06/2019     Delhi      Cochin  DEL → LKO → BOM → COK   
3       IndiGo      12/05/2019   Kolkata    Banglore        CCU → NAG → BLR   
4       IndiGo      01/03/2019  Banglore   New Delhi        BLR → NAG → DEL   

  dep_time  arrival_time duration total_stops additional_info  price  
0    22:20  01:10 22 Mar   2h 50m    non-stop         No info   3897  
1    05:50         13:15   7h 25m     2 stops         No info   7662  
2    09:25  04:25 10 Jun      19h     2 stops         No info  13882  
3    18:05         23:30   5h 25m      1 stop         No info   6218  
4    16:50         21:35   4h 45m      1 stop         No info  13302

# 3.2 Check Duplicates

In [11]:
df.duplicated().sum()

# It is showing 220 duplicate records
# So, let's check why are they duplicates

220

In [12]:
(
    df[df.duplicated(keep = False)]
    .sort_values(by = ['airline', 'date_of_journey', 'source', 'destination'])
)

airline date_of_journey    source destination                  route  \
6321  Air India      01/03/2019  Banglore   New Delhi  BLR → BOM → AMD → DEL   
9848  Air India      01/03/2019  Banglore   New Delhi  BLR → BOM → AMD → DEL   
572   Air India      03/03/2019  Banglore   New Delhi              BLR → DEL   
8168  Air India      03/03/2019  Banglore   New Delhi              BLR → DEL   
1495  Air India       1/04/2019   Kolkata    Banglore  CCU → DEL → COK → BLR   
...         ...             ...       ...         ...                    ...   
2692   SpiceJet      24/03/2019  Banglore   New Delhi              BLR → DEL   
2870   SpiceJet      24/03/2019  Banglore   New Delhi              BLR → DEL   
3711   SpiceJet      24/03/2019  Banglore   New Delhi              BLR → DEL   
2634    Vistara      24/03/2019  Banglore   New Delhi              BLR → DEL   
5470    Vistara      24/03/2019  Banglore   New Delhi              BLR → DEL   

     dep_time  arrival_time duration total_stops  \
6321    08:50  23:55 02 Mar   39h 5m     2 stops   
9848    08:50  23:55 02 Mar   39h 5m     2 stops   
572     21:10         23:55   2h 45m    non-stop   
8168    21:10         23:55   2h 45m    non-stop   
1495    10:00  01:20 02 Apr  15h 20m     2 stops   
...       ...           ...      ...         ...   
2692    05:45         08:35   2h 50m    non-stop   
2870    05:45         08:35   2h 50m    non-stop   
3711    20:30         23:20   2h 50m    non-stop   
2634    11:30         14:10   2h 40m    non-stop   
5470    11:30         14:10   2h 40m    non-stop   

                   additional_info  price  
6321                       No info  17135  
9848                       No info  17135  
572                        No info   7591  
8168                       No info   7591  
1495                       No info  10408  
...                            ...    ...  
2692  No check-in baggage included   4273  
2870  No check-in baggage included   4273  
3711  No check-in baggage included   3873  
2634                       No info   5403  
5470                       No info   5403  

[415 rows x 11 columns]

# Observations for dtype

* `date_of_journey`, `departure_time` and `arrival_time` is object but it should be datetime's object
* `duration` is like `2h 50m` convert to only minutes and dtype should be numeric.
* `total_stops` is like `2 stops` and `non stop`. Convert it to numeric dtype.

# Observation for duplicates
* There're `220` duplicate records

# 4. Detailed Analysis

In [13]:
df.head()

airline date_of_journey    source destination                  route  \
0       IndiGo      24/03/2019  Banglore   New Delhi              BLR → DEL   
1    Air India       1/05/2019   Kolkata    Banglore  CCU → IXR → BBI → BLR   
2  Jet Airways       9/06/2019     Delhi      Cochin  DEL → LKO → BOM → COK   
3       IndiGo      12/05/2019   Kolkata    Banglore        CCU → NAG → BLR   
4       IndiGo      01/03/2019  Banglore   New Delhi        BLR → NAG → DEL   

  dep_time  arrival_time duration total_stops additional_info  price  
0    22:20  01:10 22 Mar   2h 50m    non-stop         No info   3897  
1    05:50         13:15   7h 25m     2 stops         No info   7662  
2    09:25  04:25 10 Jun      19h     2 stops         No info  13882  
3    18:05         23:30   5h 25m      1 stop         No info   6218  
4    16:50         21:35   4h 45m      1 stop         No info  13302

In [14]:
df['airline']

0             IndiGo
1          Air India
2        Jet Airways
3             IndiGo
4             IndiGo
            ...     
10678       Air Asia
10679      Air India
10680    Jet Airways
10681        Vistara
10682      Air India
Name: airline, Length: 10683, dtype: object

In [15]:
df['airline'].unique()

array(['IndiGo', 'Air India', 'Jet Airways', 'SpiceJet',
       'Multiple carriers', 'GoAir', 'Vistara', 'Air Asia',
       'Vistara Premium economy', 'Jet Airways Business',
       'Multiple carriers Premium economy', 'Trujet'], dtype=object)

In [16]:
# airline

(
    df.airline
    .str.replace('Vistara Premium Economy', 'Vistara', case = False, regex = True)
    .str.replace('Jet Airways Business', 'Jet Airways', case = False, regex = True)
    .str.title()
).unique()

array(['Indigo', 'Air India', 'Jet Airways', 'Spicejet',
       'Multiple Carriers', 'Goair', 'Vistara', 'Air Asia',
       'Multiple Carriers Premium Economy', 'Trujet'], dtype=object)

In [17]:
# date_of_journey

pd.to_datetime(df['date_of_journey'], dayfirst = True)

# Pandas expects date in YYYY-MM-DD format, but our data fas DD-MM-YYYY format.
# So we are setting dayfirst = True to inform pandas that in our data, day is first

0       2019-03-24
1       2019-05-01
2       2019-06-09
3       2019-05-12
4       2019-03-01
           ...    
10678   2019-04-09
10679   2019-04-27
10680   2019-04-27
10681   2019-03-01
10682   2019-05-09
Name: date_of_journey, Length: 10683, dtype: datetime64[ns]

In [18]:
# source

df['source'].unique()

array(['Banglore', 'Kolkata', 'Delhi', 'Chennai', 'Mumbai'], dtype=object)

In [19]:
# Destination

df['destination'].unique()

array(['New Delhi', 'Banglore', 'Cochin', 'Kolkata', 'Delhi', 'Hyderabad'],
      dtype=object)

In [20]:
# Departure time

df['dep_time']

0        22:20
1        05:50
2        09:25
3        18:05
4        16:50
         ...  
10678    19:55
10679    20:45
10680    08:20
10681    11:30
10682    10:55
Name: dep_time, Length: 10683, dtype: object

In [21]:
# Getting those rows which is not following HH:MM format
# Pattern for HH:MM (24-hour)
pattern = r'^(?:[01]?\d|2[0-3]):[0-5]\d$'

# Find rows NOT matching the pattern
df[~df['dep_time'].str.match(pattern)]

# Observation: All the rows are following HH:MM pattern

Empty DataFrame
Columns: [airline, date_of_journey, source, destination, route, dep_time, arrival_time, duration, total_stops, additional_info, price]
Index: []

In [22]:
pd.to_datetime(df['dep_time'], format='%H:%M', errors='coerce').dt.time

0        22:20:00
1        05:50:00
2        09:25:00
3        18:05:00
4        16:50:00
           ...   
10678    19:55:00
10679    20:45:00
10680    08:20:00
10681    11:30:00
10682    10:55:00
Name: dep_time, Length: 10683, dtype: object

In [23]:
# Arrival time

df['arrival_time']

0        01:10 22 Mar
1               13:15
2        04:25 10 Jun
3               23:30
4               21:35
             ...     
10678           22:25
10679           23:20
10680           11:20
10681           14:10
10682           19:15
Name: arrival_time, Length: 10683, dtype: object

In [24]:
# Duration

df['duration']

0        2h 50m
1        7h 25m
2           19h
3        5h 25m
4        4h 45m
          ...  
10678    2h 30m
10679    2h 35m
10680        3h
10681    2h 40m
10682    8h 20m
Name: duration, Length: 10683, dtype: object

In [25]:
# Checking pattern missmatch

# Pattern: hours (mandatory) optionally followed by minutes
pattern = r'^\d+h(?: \d+m)?$'

# Find rows NOT matching the pattern
df[~df['duration'].str.match(pattern)]

airline date_of_journey  source destination                  route  \
6474  Air India       6/03/2019  Mumbai   Hyderabad  BOM → GOI → PNQ → HYD   

     dep_time arrival_time duration total_stops additional_info  price  
6474    16:50        16:55       5m     2 stops         No info  17327

In [26]:
# total_stops

df['total_stops']

0        non-stop
1         2 stops
2         2 stops
3          1 stop
4          1 stop
           ...   
10678    non-stop
10679    non-stop
10680    non-stop
10681    non-stop
10682     2 stops
Name: total_stops, Length: 10683, dtype: object

In [27]:
# Checking for pattern misssmatch

pattern = r'^(non-stop|\d+ stop(s)?)$'

df[~df['total_stops'].str.match(pattern, na = False)]

airline date_of_journey source destination route dep_time  \
9039  Air India       6/05/2019  Delhi      Cochin   NaN    09:45   

      arrival_time duration total_stops additional_info  price  
9039  09:25 07 May  23h 40m         NaN         No info   7480

In [28]:
# additional_info

df['additional_info'].value_counts()

# No Info and No info (Capital and small I in info)
# So replacing it

additional_info
No info                         8345
In-flight meal not included     1982
No check-in baggage included     320
1 Long layover                    19
Change airports                    7
Business class                     4
No Info                            3
1 Short layover                    1
Red-eye flight                     1
2 Long layover                     1
Name: count, dtype: int64

In [29]:
df['additional_info'].str.replace('info', 'Info').value_counts()

additional_info
No Info                         8348
In-flight meal not included     1982
No check-in baggage included     320
1 Long layover                    19
Change airports                    7
Business class                     4
1 Short layover                    1
Red-eye flight                     1
2 Long layover                     1
Name: count, dtype: int64

In [30]:
df.columns

Index(['airline', 'date_of_journey', 'source', 'destination', 'route',
       'dep_time', 'arrival_time', 'duration', 'total_stops',
       'additional_info', 'price'],
      dtype='object')

# Observation:
* There're two values indicating same airline: `Vistara` and `Vistara Premiun Economy`. Will change it to `Vistara`
* Same ffoes with `Jet Airways` and `Jet Airways Business`.
* Also there is a value `Multiple carriers Premiun economy`. We will change it to `Multiple carriers`

* Row indexed `6474` has `duration of 5 minutes` which is wrong information. So we will delete it in cleaning step.

# 5. Cleaning Operations

In [31]:
# duration: Function for converting duration column

def duration_to_minutes(s):
    h, m = 0, 0
    if 'h' in s:
        h = int(s.split('h')[0])
        if 'm' in s:
            m = int(s.split('h')[1].replace('m','').strip())
    elif 'm' in s:
        m = int(s.replace('m',''))
    return h*60 + m

In [32]:
# total_stops: Cleaning. non-stop = 0, 2 stops = 2. 3 stops = 3 and so on

def stops_to_int(x):
    if pd.isna(x):
        return float('nan')  # keep NaN
    if str(x).lower() == 'non-stop':
        return 0
    return int(str(x).split()[0])

In [33]:
def clean_data(df):

    # Stripping the values of object columns
    df = df.assign(**{col: df[col].str.strip() for col in df.select_dtypes(include='object').columns})

    # Dropping duplicate rows
    df.drop_duplicates(inplace = True)

    # Changing the columns's name to lower
    df.columns = df.columns.str.lower().str.strip()

    # Cleaning airline column
    df['airline'] = (
        df['airline']
            .str.replace('Vistara Premium Economy', 'Vistara', case = False, regex = True)
            .str.replace('Jet Airways Business', 'Jet Airways', case = False, regex = True)
            .str.replace('multiple carriers premium economy', 'Multiple Carriers', case = False, regex = True)
            .str.title()
            .str.strip()
    )

    # date_of_journey: Converting dtype of date_of_journey to datetime
    df['date_of_journey'] = pd.to_datetime(df['date_of_journey'], dayfirst = True)

    # route: Dropping route column as it is giving info from source, destination and stops.
    df.drop(columns = ['route'], inplace = True)

    # dep_time: There is no pattern missmatch in dep_time. So just converting dtype
    df['dep_time'] = pd.to_datetime(df['dep_time'], format = ("%H:%M"), errors = 'coerce').dt.time

    # arrival_time: Some values containa time in HH:MM format and some contains time + date both.
    # So extracting time from it and then converting to datetime object
    
    # Extract only HH:MM from arrival_time
    df['arrival_time'] = df['arrival_time'].str.extract(r'(\d{1,2}:\d{2})')[0]
    
    # Convert to datetime.time objects
    df['arrival_time'] = pd.to_datetime(df['arrival_time'], format='%H:%M', errors='coerce').dt.time

    # duration: converting duration into minutes
    df['duration'] = df['duration'].apply(duration_to_minutes)
    # Dropping that 6474th row having wrong duration of 5 minutes.
    df.drop(index = 6474, axis = 'row', inplace = True)

    # total_stops: Cleaning. non-stop = 0, 2 stops = 2. 3 stops = 3 and so on
    df['total_stops'] = df['total_stops'].apply(stops_to_int).astype('Int64')   # Int64 allows Nan values, else it converts to float

    # additional_info: Solving typo in additional_info
    df['additional_info'] = df['additional_info'].str.replace('info', 'Info')
    
    return df[['airline', 'date_of_journey', 'source', 'destination',
       'dep_time', 'arrival_time', 'duration', 'total_stops',
       'additional_info', 'price']]

In [34]:
df2 = clean_data(df)
df2.head()

airline date_of_journey    source destination  dep_time arrival_time  \
0       Indigo      2019-03-24  Banglore   New Delhi  22:20:00     01:10:00   
1    Air India      2019-05-01   Kolkata    Banglore  05:50:00     13:15:00   
2  Jet Airways      2019-06-09     Delhi      Cochin  09:25:00     04:25:00   
3       Indigo      2019-05-12   Kolkata    Banglore  18:05:00     23:30:00   
4       Indigo      2019-03-01  Banglore   New Delhi  16:50:00     21:35:00   

   duration  total_stops additional_info  price  
0       170            0         No Info   3897  
1       445            2         No Info   7662  
2      1140            2         No Info  13882  
3       325            1         No Info   6218  
4       285            1         No Info  13302

# 6. Splitting the data into
* Train
* Test
* Validation

In [35]:
sample_df = df2.sample(frac = 1, random_state = 1).reset_index(drop = True)

In [36]:
x = sample_df.drop(columns = 'price')
y = sample_df.price.copy()

In [37]:
# Keeping 15% for testing
x_train, x_temp, y_train, y_temp = train_test_split(x, y, test_size = 0.3, random_state = 1)

# Step 2: Split x_temp / y_temp into test & validation
x_val, x_test, y_val, y_test = train_test_split(
    x_temp, y_temp, test_size = 0.5, random_state = 1
)

# Ignore x_temp and y_temp later; they’re just temporary.
# Only the final splits matter for training, validation, and testing.

In [38]:
print(f"""Training data: {x_train.shape}, {y_train.shape},
Validation data: {x_val.shape}, {y_val.shape},
Testing data: {x_test.shape}, {y_test.shape}
""")

Training data: (7323, 9), (7323,),
Validation data: (1569, 9), (1569,),
Testing data: (1570, 9), (1570,)



# 7. Exporting subsets

In [39]:
data_dir = r"D:\Nikss\Projects\Flight Price Prediction\Datasets"

In [40]:
def export_subset(x, y, name):
	file_name = f"{name}.csv"
	file_path = os.path.join(data_dir, file_name)

    # Join the target variable and save to csv 
	x.join(y).to_csv(file_path, index=False)

	return pd.read_csv(file_path).head()

In [41]:
subsets = [
    (x_train, y_train, 'train_data'),
    (x_val, y_val, 'validation_data'),
    (x_test, y_test, 'test_data')
]

In [42]:
for subset in subsets:
    export_subset(*subset)
    print(f"Subset created")

Subset created
Subset created
Subset created


# `Next Session`: EDA